cl/opinion_page/utils.py

```diff
@@ -162,3 +133,215 @@ def generate_docket_entries_csv_data(docket_entries):
     csv_content: str = output.getvalue()
     output.close()
     return csv_content
+
+
+async def build_cites_clusters_query(
+    cluster_search: Search, sub_opinion_pks: list[str]
+) -> Search:
+    """Build the ES cites clusters query to find clusters citing specific
+    sub-opinions.
+
+    :param cluster_search: The Elasticsearch DSL Search object
+    :param sub_opinion_pks: A list of ids of sub-opinions to search for in
+    the 'cites' field.
+    :return: The ES DSL Search object representing the query to find clusters
+    citing the specified sub-opinions.
+    """
+    cites_query = Q(
+        "bool",
+        filter=[
+            Q("match", cluster_child="opinion"),
+            Q("terms", **{"cites": sub_opinion_pks}),
+        ],
+    )
+    cluster_cites_query = cluster_search.query(cites_query)
+    search_query = (
+        cluster_cites_query.sort({"citeCount": {"order": "desc"}})
+        .source(includes=["absolute_url", "caseName", "dateFiled"])
+        .extra(size=5, track_total_hits=True)
+    )
+    return search_query
+
+
+async def build_related_clusters_query(
+    cluster_search: Search,
+    sub_opinion_pks: list[str],
+    search_params: dict[str, str],
+) -> Search:
+    """Build the ES related clusters query based on sub-opinion IDs.
+
+    :param cluster_search: The Elasticsearch DSL Search object
+    :param sub_opinion_pks: A list of IDs representing sub-opinions to be queried.
+    :param search_params: A dict of parameters used to form the query.
+    :return: The ES DSL Search object representing the query to find the
+    related clusters.
+    """
+    mlt_query = await build_more_like_this_query(sub_opinion_pks)
+    parent_filters = await sync_to_async(build_join_es_filters)(
+        {"type": SEARCH_TYPES.OPINION, "stat_published": True}
+    )
+    default_parent_filter = [Q("match", cluster_child="opinion")]
+    parent_filters.extend(default_parent_filter)
+    main_query = Q(
+        "bool",
+        filter=default_parent_filter,
+        should=mlt_query,
+        minimum_should_match=1,
+    )
+
+    cluster_related_query = cluster_search.query(main_query)
+    search_query = (
+        cluster_related_query.sort({"_score": {"order": "desc"}})
+        .source(includes=["absolute_url", "caseName", "cluster_id"])
+        .extra(size=5)
+        .collapse(field="cluster_id")
+    )
+    return search_query
+
+
+async def es_get_citing_and_related_clusters_with_cache(
+    cluster: OpinionCluster,
+    request: HttpRequest,
+) -> tuple[
+    list[OpinionClusterDocument],
+    list[int],
+    dict[str, str],
+    list[OpinionClusterDocument],
+    int,
+    bool,
+]:
```

Comment:
I believe that using a data class instead of a tuple can make this code easier to read and understand. Tuples can get confusing when you have a lot of things in them, especially if you need to remember the right order. If you need to add something new to the tuple later, it can be a pain to change everything else. With a data class, it’s much simpler and less error-prone.

Reply:
Yeah the dataclass is a good fit here. I've added it (RelatedCitingResults)